array_api_compat/torch/_aliases.py

```diff
@@ -0,0 +1,560 @@
+    if not isinstance(from_, torch.dtype):
+        from_ = from_.dtype
+    return torch.can_cast(from_, to)
+
+# Basic renames
+permute_dims = torch.permute
+bitwise_invert = torch.bitwise_not
+
+# Two-arg elementwise functions
+# These require a wrapper to do the correct type promotion on 0-D tensors
+add = _two_arg(torch.add)
+atan2 = _two_arg(torch.atan2)
+bitwise_and = _two_arg(torch.bitwise_and)
+bitwise_left_shift = _two_arg(torch.bitwise_left_shift)
+bitwise_or = _two_arg(torch.bitwise_or)
+bitwise_right_shift = _two_arg(torch.bitwise_right_shift)
+bitwise_xor = _two_arg(torch.bitwise_xor)
+divide = _two_arg(torch.divide)
+# Also a rename. torch.equal does not broadcast
+equal = _two_arg(torch.eq)
+floor_divide = _two_arg(torch.floor_divide)
+greater = _two_arg(torch.greater)
+greater_equal = _two_arg(torch.greater_equal)
+less = _two_arg(torch.less)
+less_equal = _two_arg(torch.less_equal)
+logaddexp = _two_arg(torch.logaddexp)
+# logical functions are not included here because they only accept bool in the
+# spec, so type promotion is irrelevant.
+multiply = _two_arg(torch.multiply)
+not_equal = _two_arg(torch.not_equal)
+pow = _two_arg(torch.pow)
+remainder = _two_arg(torch.remainder)
+subtract = _two_arg(torch.subtract)
+
+# These wrappers are mostly based on the fact that pytorch uses 'dim' instead
+# of 'axis'.
+
+# torch.min and torch.max return a tuple and don't support multiple axes https://github.com/pytorch/pytorch/issues/58745
+def max(x: array, /, *, axis: Optional[Union[int, Tuple[int, ...]]] = None, keepdims: bool = False) -> array:
+    # https://github.com/pytorch/pytorch/issues/29137
+    if axis == ():
+        return torch.clone(x)
+    return torch.amax(x, axis, keepdims=keepdims)
+
+def min(x: array, /, *, axis: Optional[Union[int, Tuple[int, ...]]] = None, keepdims: bool = False) -> array:
+    # https://github.com/pytorch/pytorch/issues/29137
+    if axis == ():
+        return torch.clone(x)
+    return torch.amin(x, axis, keepdims=keepdims)
+
+# torch.sort also returns a tuple
+# https://github.com/pytorch/pytorch/issues/70921
+def sort(x: array, /, *, axis: int = -1, descending: bool = False, stable: bool = True, **kwargs) -> array:
+    return torch.sort(x, dim=axis, descending=descending, stable=stable, **kwargs).values
+
+def _normalize_axes(axis, ndim):
+    axes = []
+    lower, upper = -ndim, ndim - 1
+    for a in axis:
+        if a < lower or a > upper:
+            # Match torch error message (e.g., from sum())
+            raise IndexError(f"Dimension out of range (expected to be in range of [{lower}, {upper}], but got {a}")
+        if a < 0:
+            a = a + ndim
+        if a in axes:
+            # Match torch error message but use IndexError instead of RuntimeError
+            raise IndexError(f"dim {a} appears multiple times in the list of dims")
+        axes.append(a)
+    return sorted(axes)
+
+
+def _apply_keepdims(x, ndim, keepdims):
+    if keepdims:
+        return x[(None,)*ndim]
```

Comment:
This seems wrong. You are not passing the sizes of the of the initial vector, so how do you know what's the shape the tensor `x` needs to have. This should be done in PyTorch via `unsqueeze`s and a `expand`. See for example:
https://github.com/pytorch/pytorch/blob/f3bf46e801dec2637751224fd6e27fbf97453bc6/torch/_refs/__init__.py#L2057-L2060
and
https://github.com/pytorch/pytorch/blob/f3bf46e801dec2637751224fd6e27fbf97453bc6/torch/_prims/__init__.py#L1246-L1256

Reply:
Just need to rename the helper here. I only need to manually apply keepdims when `axis=None`, because of https://github.com/pytorch/pytorch/issues/71209. So it's really just a special case of expanding a 0-D tensor to a (1, 1, ..., 1) tensor. If there's a better way to do it than `newaxis` indexing I can do that instead.

Reply:
Renaming and a comment would be good, yep.
And yes, this is best done in PyTorch via a number of `unsqueeze`s

Reply:
I noticed this internal function in pytorch that does exactly this, although it isn't using either of these methods https://github.com/pytorch/pytorch/blob/e61d5b95886c11340118f559529397a7d0556d6b/torch/_prims/__init__.py#L1411.

Reply:
It does use `unsqueeze` and `expand` at the end, as it calls into `broadcast_in_dim`:
https://github.com/pytorch/pytorch/blob/f3bf46e801dec2637751224fd6e27fbf97453bc6/torch/_prims/__init__.py#L1246-L1256